lib/processExportsField.js

```diff
@@ -0,0 +1,270 @@
+	}
+
+	const isFolder = exp.charCodeAt(exp.length - 1) === slashCode;
+
+	if (isFolder !== expectFolder) {
+		throw new Error(
+			expectFolder
+				? "Expecting folder to folder mapping. Folder to file found"
+				: "Expecting file to file mapping. File to folder found"
+		);
+	}
+}
+
+/**
+ * @param {ConditionalMapping|Mapping} mapping mapping
+ * @returns {boolean} is conditional mapping
+ */
+function isConditionalMapping(mapping) {
+	return typeof mapping === "object" && !Array.isArray(mapping);
+}
+
+/**
+ * @param {string} remainingRequest remaining request
+ * @param {boolean} expectFolder expect folder
+ * @param {string|string[]} directExport direct export
+ * @returns {string[]} mapping result
+ */
+function directMapping(remainingRequest, expectFolder, directExport) {
+	const isString = typeof directExport === "string";
+
+	if (isString) {
+		assertExport(/** @type {string} */ (directExport), expectFolder);
+	} else {
+		/** @type {string[]} */ (directExport).forEach(exp =>
+			assertExport(exp, expectFolder)
+		);
+	}
+
+	if (!expectFolder) {
+		return isString
+			? [/** @type {string} */ (directExport)]
+			: /** @type {string[]} */ (directExport).slice();
+	}
+
+	return isString
+		? [/** @type {string} */ (directExport.concat(remainingRequest))]
+		: /** @type {string[]} */ (directExport).map(d =>
+				d.concat(remainingRequest)
+		  );
+}
+
+/**
+ * @param {string} remainingRequest remaining request
+ * @param {boolean} expectFolder expect folder
+ * @param {ConditionalMapping} conditionalExport direct export
+ * @param {Set<string>} conditionNames condition names
+ * @returns {string[]} mapping result
+ */
+function conditionalMapping(
+	remainingRequest,
+	expectFolder,
+	conditionalExport,
+	conditionNames
+) {
+	/** @type {ConditionalMapping|null} */
+	let nextLookup = conditionalExport;
+
+	loop: while (nextLookup !== null) {
+		const conditions = Object.keys(nextLookup);
+		const last = conditions.length - 1;
+
+		for (let i = 0; i < conditions.length; i++) {
+			const condition = conditions[i];
+
+			// assert default. Could be last only
+			if (i !== last) {
+				if (condition === "default") {
+					throw new Error("Default condition should be last one");
+				}
+			} else if (condition === "default") {
+				const mapping = nextLookup[condition];
+				// is nested
+				if (isConditionalMapping(mapping)) {
+					nextLookup = mapping;
+					continue loop;
+				}
+
+				return directMapping(
+					remainingRequest,
+					expectFolder,
+					/** @type {string|string[]} */ (mapping)
+				);
+			}
+
+			if (conditionNames.has(condition)) {
+				const mapping = nextLookup[condition];
+				// is nested
+				if (isConditionalMapping(mapping)) {
+					nextLookup = /** @type {ConditionalMapping} */ (mapping);
+					continue loop;
+				}
+
+				return directMapping(
+					remainingRequest,
+					expectFolder,
+					/** @type {string|string[]} */ (mapping)
+				);
+			}
+		}
+
+		nextLookup = null;
```

Comment:
``` json
{
  "exports": {
    "abc": { "def": "./x.js" },
    "ghi": "./y.js"
  }
}
```

With `abc` and `ghi` being active conditions, while `def` is not set.

from https://nodejs.org/docs/latest-v13.x/api/esm.html#esm_nested_conditions

> If a nested conditional does not have any mapping it will continue checking the remaining conditions of the parent condition.

The result should be `["./y.js"]`.

Either use a lookup stack, or use recursion. I would go with recursion, as it's easier to implement here.

Another edge case test case:

``` json
{
  "exports": {
    "abc": { "def": "./x.js", "default": [] },
    "ghi": "./y.js"
  }
}
```

should result in `[]` with `abc` and `ghi`